### PR TITLE
Update of multisegment well pressure equations - testing

### PR DESCRIPTION
**Pull request by steink**
* rewrite segment hydrostatic p-drop as average of neighboring nodes
* slightly hacky inclusion of "skipped" derivatives in the friction-term for reverse-flow cases. Could be done similarly also for valve-models, but currently not included for these.

### CODE REVIEW

**Comment by GitPaean** [comment]
benchmark please

**Comment by GitPaean** [comment]
jenkins build this please

**Comment by ytelses** [comment]
Benchmark result overview: 




| Test    | Configuration                              | Relative |
| ------- | ------------------------------------------ | -------- |
| opm-git | OPM Benchmark: drogon - Threads: 1         | 1.059    |
| opm-git | OPM Benchmark: drogon - Threads: 8         | 1.393    |
| opm-git | OPM Benchmark: spe10_model_1 - Threads: 1  | 0.995    |
| opm-git | OPM Benchmark: spe10_model_1 - Threads: 8  | 1.007    |
| opm-git | OPM Benchmark: flow_mpi_extra - Threads: 1 | 1.022    |
| opm-git | OPM Benchmark: flow_mpi_extra - Threads: 8 | 1.028    |
| opm-git | OPM Benchmark: flow_mpi_norne - Threads: 1 | 0.999    |
| opm-git | OPM Benchmark: flow_mpi_norne - Threads: 8 | 0.984    |
* Speed-up = Total time master / Total time pull request. Above 1.0 is an improvement. *



View result details @ https://www.ytelses.com/opm/?page=result&id=2073

File: opm/simulators/wells/MultisegmentWellEval.cpp [hunk 1]
```diff
@@ -217,22 +217,27 @@ assembleDefaultPressureEq(const int seg,
...
     auto& segments = ws.segments;
-    segments.pressure_drop_hydrostatic[seg] = hydro_pressure_drop.value();
-    pressure_equation -= hydro_pressure_drop;
+    //pressure_equation -= hydro_pressure_drop;
 
     if (this->frictionalPressureLossConsidered()) {
-        const auto friction_pressure_drop = segments_.getFrictionPressureLoss(seg);
+        const auto friction_pressure_drop = segments_.getFrictionPressureLoss(seg, false);
+        if (reverseFlow){
+            extra_derivatives = -segments_.getFrictionPressureLoss(seg, true);
```
Comment (GitPaean) [inline opm/simulators/wells/MultisegmentWellEval.cpp:239]: The following code should go into this bracket? or at least inside the bracket (`if (this->frictionalPressureLossConsidered()) {`)
```c++
        MultisegmentWellAssemble<FluidSystem,Indices,Scalar>(baseif_).
            assemblePressureEqExtraDerivatives(seg, seg_upwind, extra_derivatives, linSys_);
```

**Review by GitPaean — CHANGES_REQUESTED** [review]
This generally looks fine. The comments are very minor.  I need some more time to think about the frictional pressure loss part.

File: opm/simulators/wells/MultisegmentWellEval.cpp (hunk 1 above)
Comment (GitPaean) [reply opm/simulators/wells/MultisegmentWellEval.cpp:239]: In the current form, if there is no frictional pressure loss, calling the function `assemblePressureEqExtraDerivatives` can be dangerous?

File: opm/simulators/wells/MultisegmentWellSegments.cpp [hunk 2]
```diff
@@ -484,19 +487,36 @@ getSurfaceVolume(const EvalWell& temperature,
...
+            const int WQTotal = Indices::numEq + PrimaryVariables::WQTotal;
+            const int SPres = Indices::numEq + PrimaryVariables::SPres;
+            density.setDerivative(WQTotal, 0.0);
+            density.setDerivative(SPres, 0.0);
+            visc.setDerivative(WQTotal, 0.0);
+            visc.setDerivative(SPres, 0.0);
+        } else {
+            const int WFrac = Indices::numEq + PrimaryVariables::WFrac;
+            const int GFrac = Indices::numEq + PrimaryVariables::GFrac;
+            // if has ...
```
Comment (GitPaean) [inline opm/simulators/wells/MultisegmentWellSegments.cpp:511]: Yeah, we need to have the guard in case there is no water or gas involved.

**Comment by GitPaean** [comment]
And also, can you rebase to the latest branch? It is not required while recommended.

File: opm/simulators/wells/MultisegmentWellSegments.cpp [hunk 3]
```diff
@@ -321,11 +321,14 @@ updateUpwindingSegments(const PrimaryVariables& primary_variables)
 template<class FluidSystem, class Indices, class Scalar>
 typename MultisegmentWellSegments<FluidSystem,Indices,Scalar>::EvalWell
 MultisegmentWellSegments<FluidSystem,Indices,Scalar>::
-getHydroPressureLoss(const int seg) const
-{
-    return densities_[seg] * well_.gravity() * depth_diffs_[seg];
+getHalfHydroPressureLoss(const int seg,
+                         const int seg_side) const
```
[the remark this answers is not in the document]
Comment (steink) [reply opm/simulators/wells/MultisegmentWellSegments.cpp]: Done

File: opm/simulators/wells/MultisegmentWellEval.cpp [hunk 4]
```diff
@@ -217,22 +217,27 @@ assembleDefaultPressureEq(const int seg,
...
     // for top segment, the well control equation will be used.
     EvalWell pressure_equation = primary_variables_.getSegmentPressure(seg);
+    EvalWell extra_derivatives;
 
     // we need to handle the pressure difference between the two segments
     // we only consider the hydrostatic pressure loss first
     // TODO: we might be able to add member variables to store these values, then we update well state
     // after converged
-    const auto hydro_pressure_drop = segments_.getHydroPressureLoss(seg);
+
```
[the remark this answers is not in the document]
Comment (steink) [reply opm/simulators/wells/MultisegmentWellEval.cpp:231]: Done

File: opm/simulators/wells/MultisegmentWellEval.cpp (hunk 1 above)
Comment (steink) [reply opm/simulators/wells/MultisegmentWellEval.cpp:249]: Yes, moved `assemblePressureEqExtraDerivatives` up into the `(if (this->frictionalPressureLossConsidered())  `-bracket

File: opm/simulators/wells/MultisegmentWellSegments.cpp [hunk 5]
```diff
@@ -484,19 +487,36 @@ getSurfaceVolume(const EvalWell& temperature,
...
     EvalWell visc = viscosities_[seg_upwind];
     // WARNING
     // We disregard the derivatives from the upwind density to make sure derivatives
     // wrt. to different segments dont get mixed.
+    
     if (seg != seg_upwind) {
-        density.clearDerivatives();
-        visc.clearDerivatives();
+        if (!return_upwind_derivatives){
+            const int WQTotal = Indices::numEq + PrimaryVariables::WQTotal;
```
[the remark this answers is not in the document]
Comment (steink) [reply opm/simulators/wells/MultisegmentWellSegments.cpp]: Done

**Comment by steink** [comment]
After the update 6dc90b2df77e4bdfc0af9d0e8d548f66bdf96bbf the behavior has slightly changed (one time-step cut versus zero for a simple test case). This is not really expected as the only functional change is the order of which the various terms of the MS pressure equations are assembled.

**Comment by GitPaean** [comment]
> After the update [6dc90b2](https://github.com/OPM/opm-simulators/commit/6dc90b2df77e4bdfc0af9d0e8d548f66bdf96bbf) the behavior has slightly changed (one time-step cut versus zero for a simple test case). This is not really expected as the only functional change is the order of which the various terms of the MS pressure equations are assembled.

hei, at the end, it is the different orders of the assembly causing some deviations during the floating point arithmetic operations, which causes the simulations deviates a little bit. 

To recover the running results before commit https://github.com/OPM/opm-simulators/commit/6dc90b2df77e4bdfc0af9d0e8d548f66bdf96bbf , you can move the following code to the end of the function. 

```c++
    // We approximate the hydrostatic pressure loss over the segment by using the mean of the densities
    // at the current segment node and its outlet node. Since density derivatives are organized
    // differently than what is required for assemblePressureEq, this part needs to be assembled separately.
    const int seg_outlet = this->segmentNumberToIndex(this->segmentSet()[seg].outletSegment());
    const auto hydro_pressure_drop_seg = segments_.getHalfHydroPressureLoss(seg, seg);
    const auto hydro_pressure_drop_outlet = segments_.getHalfHydroPressureLoss(seg, seg_outlet);
    MultisegmentWellAssemble<FluidSystem,Indices,Scalar>(baseif_).
            assembleHydroPressureLoss(seg, seg_outlet, hydro_pressure_drop_seg, hydro_pressure_drop_outlet, linSys_);
    segments.pressure_drop_hydrostatic[seg] = hydro_pressure_drop_seg.value() + hydro_pressure_drop_outlet.value();
```

**Comment by GitPaean** [comment]
jenkins build this please

**Comment by steink** [comment]
> To recover the running results before commit https://github.com/OPM/opm-simulators/commit/6dc90b2df77e4bdfc0af9d0e8d548f66bdf96bbf , you can move the following code to the end of the function.

Yes, this recovers the results. Done.

> Personally, I think the following function should be called the first during the pressureEq assembly,

I kind of agree, but kept it as is for now for compactness. I guess we will want to return here at a later stage to check whether we should have special treatment of acceleration derivatives as well (possibly by reorganizing the assembling).

**Comment by steink** [comment]
Reverted segment density calculations to the original formulation since this seemed to agree more closely with the reference. However, kept the average formulation as a hidden option (_--use-average-density-ms-wells_) since it appears to be more stable and more in line with the analogous formulation for the reservoir grid cells.

**Comment by GitPaean** [comment]
jenkins build this please

**Comment by GitPaean** [comment]
benchmark please

**Comment by GitPaean** [comment]
I am checking the jenkins failures now before preparing to get the PR in.

**Comment by ytelses** [comment]
Benchmark result overview: 




| Test    | Configuration                              | Relative |
| ------- | ------------------------------------------ | -------- |
| opm-git | OPM Benchmark: drogon - Threads: 1         | 0.764    |
| opm-git | OPM Benchmark: drogon - Threads: 8         | 1.098    |
| opm-git | OPM Benchmark: smeaheia - Threads: 1       | 1.024    |
| opm-git | OPM Benchmark: smeaheia - Threads: 8       | 1.013    |
| opm-git | OPM Benchmark: spe10_model_1 - Threads: 1  | 1.015    |
| opm-git | OPM Benchmark: spe10_model_1 - Threads: 8  | 1.008    |
| opm-git | OPM Benchmark: flow_mpi_extra - Threads: 1 | 1.043    |
| opm-git | OPM Benchmark: flow_mpi_extra - Threads: 8 | 1.021    |
| opm-git | OPM Benchmark: flow_mpi_norne - Threads: 1 | 1.031    |
| opm-git | OPM Benchmark: flow_mpi_norne - Threads: 8 | 1.013    |
* Speed-up = Total time master / Total time pull request. Above 1.0 is an improvement. *



View result details @ https://www.ytelses.com/opm/?page=result&id=2091

**Comment by GitPaean** [comment]
I reproduced the following benchmark results locally. 



Test | Configuration | Relative
-- | -- | --
opm-git | OPM Benchmark: drogon - Threads: 1 | 0.764
opm-git | OPM Benchmark: drogon - Threads: 8 | 1.098


Will discuss and evaluate.

**Comment by steink** [comment]
A comment on **OPM Benchmark: drogon - Threads: 1** where there is a perfomance drop with the PR:

- Difference in run-time is mainly due to step 113 where there is a change in valve config. At this step, a reference run (with small time steps) shows an abrupt increase in WOPR (see fig).
- All configurations I ran (both master and PR) chops time-steps at this stage and captures the narrow peak. The only exception is master serial with default options which jumps over the peak. It also has convergence problems here, but accepts the step with relaxed tolerance.
- The behaviour of master serial at this step is very sensitive to perturbations. For instance, if it is restarted from HIST of some of  the other configurations, it also does the chopping at this stage.      

image

**Comment by GitPaean** [comment]
jenkins build this please

**Comment by bska** [comment]
> A comment on **OPM Benchmark: drogon - Threads: 1**

Thank you so much for the detailed analysis, @steink!  This is greatly appreciated.  Computing a more accurate numerical solution is of course vastly better than computing an incorrect solution so I think we should get this into the master branch rather quickly.

**Comment by GitPaean** [comment]
This PR does not change (by default) the physical equations we are solving, but it updates/improves the derivatives and also the jacobian matrix we use, which affects the convergence behavoir. 

I checked the all the jenkins failures, most of the deviations are from different time stepping, which results from different convergence behavoir.  Some cases are more sensitive to the time stepping changes than other cases.

The performance degeneration (less than 25%) of the drogon case is also one of them.  And also, for the drogon case, the performance degeneration of the serial running are related to introduction of WSEGVALV keyword. It is possible when we improve the derivative regarding the inflow control devices, the situation might change. At the same time, in benchmarking/locall test, we also see 10% performance gain with the same code for drogon case in the parallel running. 

I think this PR is ready to update the reference data and prepare to be merged. 

I will try to do some refactoring after this PR get merged, hopefully for better readability and easier development.

**Review by GitPaean — APPROVED** [review]

**Comment by GitPaean** [comment]
I am updating the reference data to prepare to get this PR in.

**Comment by GitPaean** [comment]
jenkins build this update_data please

**Comment by GitPaean** [comment]
jenkins build this opm-tests=978 please

**Comment by GitPaean** [comment]
The reference data is updated. I am getting the PR in now.

**Pull request merged**